### ISSUE TITLE
CA3075 DoNotUseInsecureDTDProcessing false negative

### ISSUE THREAD

**Issue by IlyaGrebnov** [dotnet/roslyn-analyzers#1591]
#### Analyzer package

I am using latest VS2017 version 15.5.7 with Microsoft.NetFramework.Analyzers, version 2.6.0

#### Analyzer

CA3075 DoNotUseInsecureDTDProcessing 

#### Repro steps

        /// <summary>
        /// Parses the XML extended error.
        /// </summary>
        /// <param name="textReader">The text reader.</param>
        private static StorageExtendedErrorInformation ParseXmlExtendedError(TextReader textReader)
        {
            var nameTable = new NameTable();
            var codeElement = nameTable.Add("Code");
            var messageElement = nameTable.Add("Message");

            using (var xmlReader = XmlReader.Create(textReader, new XmlReaderSettings { NameTable = nameTable, DtdProcessing = DtdProcessing.Prohibit, XmlResolver = null }))
            {
                var errorInformation = new StorageExtendedErrorInformation();

                while (xmlReader.Read())
                {
                    if (xmlReader.NodeType == XmlNodeType.Element && object.ReferenceEquals(xmlReader.LocalName, codeElement))
                    {
                        ReflectionMethods.SetErrorCode(errorInformation, xmlReader.ReadElementContentAsString());

                        if (xmlReader.NodeType == XmlNodeType.Element && object.ReferenceEquals(xmlReader.LocalName, messageElement))
                        {
                            ReflectionMethods.SetErrorMessage(errorInformation, xmlReader.ReadElementContentAsString());
                        }
                    }
                    else if (xmlReader.NodeType == XmlNodeType.Element && object.ReferenceEquals(xmlReader.LocalName, messageElement))
                    {
                        ReflectionMethods.SetErrorMessage(errorInformation, xmlReader.ReadElementContentAsString());

                        if (xmlReader.NodeType == XmlNodeType.Element && object.ReferenceEquals(xmlReader.LocalName, codeElement))
                        {
                            ReflectionMethods.SetErrorCode(errorInformation, xmlReader.ReadElementContentAsString());
                        }
                    }
                }

                return errorInformation;
            }
        }

#### Expected behavior
No violation is reported.

#### Actual behavior
Violation for CA3075 DoNotUseInsecureDTDProcessing is reported.

**Comment by mavasani** [comment]
Also tagging @ivanbasov, who has looked at this analyzer in the past. I think we need to use flow analysis in this analyzer to get more precise results.

**Comment by jnm2** [comment]
I haven't figured out how to call `XmlReader.Create` with any `new XmlWriterSettings { DtdProcessing = }` argument. Is there a workaround besides disabling the rule?

**Comment by jnm2** [comment]
`DtdProcessing.Prohibit` documentation says that it is the default behavior, so `new XmlWriterSettings { }` is secure by default, right? And if you do manually specify `DtdProcessing`, I wouldn't expect the analyzer to flag it since you've clearly stated your intent?

**Comment by chuchu** [comment]
Is the described behavior fixed with the referenced merge?

**Comment by dotpaul** [comment]
No, #1838 is unrelated and doesn't fix the behavior described here.

**Comment by chuchu** [comment]
In the code for the analyzer CA3075 I see that it checks if the used framework version can be considered as secure. But even when a secure version is used the analyzer reports a finding as mentioned in #1880 which was closed as duplicate of this issue.
Why does the check against the framework version does not work? Is because of the missing flow analysis?
I would like to avoid the workaround because we have a lot of findings.

**Comment by dotpaul** [comment]
There are a lot of scenarios in CA3075, so it depends on the specific case.  Would you mind sharing some of the example code?

**Comment by dotpaul** [comment]
That does seem like a false positive, and I don't think a fix requires dataflow analysis for this case.  I'll see if I can fix non-DFA cases later this month.  (Or if anyone wants to get to it before me, https://www.owasp.org/index.php/XML_External_Entity_(XXE)_Prevention_Cheat_Sheet#.NET has pretty good guidance on what the analyzers should report on.)

**Comment by dotpaul** [comment]
Hi @chuchu, I've been looking more into this, and you still may want to disable DTD processing by setting DtdProcessing = DtdProcessing.Prohibit or DtdProcessing.Ignore.  Otherwise untrusted input can be an "XML bomb".  See https://msdn.microsoft.com/en-us/magazine/ee335713.aspx for an example.

These days, there seems to be a limit by default on local entity expansion (10 MB?), but parsing still may be an expensive operation, making denial of service attacks easier.

**Comment by jnm2** [comment]
`Prohibit` is the default setting though, isn't it? So specifying it wouldn't change anything.

**Comment by dotpaul** [comment]
Hi @jnm2, I haven't seen that in my testing.  For example, this will print out "haha":

```csharp
class Program
{
    static void Main(string[] args)
    {
        string hahaXml = @"<?xml version=""1.0""?>
<!DOCTYPE lolz [
<!ENTITY ha ""ha"">
<!ENTITY ha2 ""&ha;&ha;"">
]>
<lolz> &ha2; </lolz>";
        XmlTextReader x = new XmlTextReader(new StringReader(hahaXml));
        while (x.Read())
        {
            if (x.NodeType == XmlNodeType.Element)
            {
                Console.WriteLine(x.ReadElementContentAsString());
            }
        }
    }
}
```

**Comment by dotpaul** [comment]
XmlReader.Create with its defaults looks more like a false positive.  I'll try to validate more later this week, and open a new issue if needed.
